File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/CatalogAdapter.java
package pl.tysia.maggstone.ui.presentation_logic.adapter;


import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collection;

import pl.tysia.maggstone.ui.presentation_logic.filterer.CatalogFilterer;
import pl.tysia.maggstone.ui.presentation_logic.filterer.Filterer;


public abstract class CatalogAdapter<T extends ICatalogable, H extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<H> {
    protected ArrayList<T> shownItems;
    public ArrayList<T> allItems;

    protected Filterer filterer;

    protected T selectedItem;

    private ArrayList<ItemSelectedListener> listeners;
    protected ArrayList<EmptyListListener> emptyListeners;


    public T getSelectedItem() {
        return selectedItem;
    }

    public Filterer getFilterer() {
        return filterer;
    }

    public void setFilterer(Filterer filterer) {
        this.filterer = filterer;
    }

    protected void onItemClick(View v, int adapterPosition){
        T item = shownItems.get(adapterPosition);

        selectedItem = item;
        notifyDataSetChanged();

        for (ItemSelectedListener listener : listeners) listener.onItemSelected(item);
    }

    public void addItemSelectedListener(ItemSelectedListener listener){
        listeners.add(listener);
    }

    public void addEmptyListener(EmptyListListener listener){
        emptyListeners.add(listener);
    }

    public void addItem(T item){
        allItems.add(0, item);
    }

    public void addAll(Collection<T> items){
        allItems.addAll(items);
    }


    public CatalogAdapter(ArrayList<T> items) {
        allItems = items;
        shownItems = new ArrayList<>();

        filterer = new CatalogFilterer<>(allItems, shownItems);

        listeners = new ArrayList<>();
        emptyListeners = new ArrayList<>();

    }

    public void filter(){
        filterer.filter();
    }


    @Override
    public int getItemCount() {
        return shownItems.size();
    }

    public interface ItemSelectedListener<T extends ICatalogable>{
        void onItemSelected(T item);
    }

    public interface EmptyListListener{
        void onListEmptied();
    }


}

File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/CatalogAdapter.kt
package pl.tysia.maggstone.ui.presentation_logic.adapter

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import pl.tysia.maggstone.ui.presentation_logic.filterer.CatalogFilterer
import pl.tysia.maggstone.ui.presentation_logic.filterer.Filterer
import java.util.*

abstract class CatalogAdapter<T : ICatalogable, H : RecyclerView.ViewHolder>(var allItems: ArrayList<T>) :
    RecyclerView.Adapter<H>() {
    var filterer: Filterer<T>

    var shownItems: ArrayList<T> = ArrayList()
        protected set

    open var selectedItem: T? = null

    val listeners: ArrayList<ItemSelectedListener<T>>
    protected var changeListeners: ArrayList<ListChangeListener>

    protected open fun onItemClick(v: View?, adapterPosition: Int) {
        val item = shownItems[adapterPosition]
        selectedItem = item
        notifyDataSetChanged()
        for (listener in listeners) listener.onItemSelected(item)
    }

    fun addItemSelectedListener(listener: ItemSelectedListener<T>) {
        listeners.add(listener)
    }

    fun addChangeListener(listener: ListChangeListener) {
        changeListeners.add(listener)
    }

    open fun addItem(item: T) {
        allItems.add(0, item)
    }

    fun addAll(items: Collection<T>?) {
        allItems.addAll(items!!)
    }

    fun filter() {
        filterer.filter()
    }

    override fun getItemCount(): Int {
        return shownItems.size
    }

    interface ItemSelectedListener<T : ICatalogable?> {
        fun onItemSelected(item: T)
    }

    interface ListChangeListener {
        fun onListChanged()
    }

    init {
        filterer = CatalogFilterer<T>(allItems, shownItems)
        listeners = ArrayList()
        changeListeners = ArrayList()
    }
}